runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.flink.metrics;
+
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.PrintStream;
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.metrics.MetricConfig;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.reporter.AbstractReporter;
+
+/**
+ * Flink {@link org.apache.flink.metrics.reporter.MetricReporter metrics reporter} for writing
+ * metrics to a file (specified via the "metrics.reporter.test.file" config key).
+ */
+public class FileReporter extends AbstractReporter {
```

Comment:
I originally gave it a name like that, however I talked myself in to that this is a potentially more general-purpose abstraction, and it is Flink specific, so that this (`main`, not `test`) package is an OK place for it.

If it was `TestFileReporter` then I'd feel obliged to put it in the test package, at which point the build plumbing to start a job-server in `flink_runner_test.py` that has the :beam-runners-flink` `-tests.jar` on the classpath would become more convoluted, so that was an added incentive to view this as not a "tests only" thing 😀

lmk what you think!

Reply:
I get why it's not practical to have it in the test scope. That's fine. Still, I think `TestFileReporter` reflects more accurately what it does. Also, it's configured via a test property, so I don't see how it is useful for non-testing.

Reply:
> it's configured via a test property

ah, [I left the reporter alias as `test` in `flink_runner_test`](https://github.com/apache/beam/pull/7183/files#diff-195e4a1918aec54f6245e49081cadb5aR99) when I renamed this class to remove the `Test` prefix. I should change that alias to `file`, either way.

This is a reporter that writes the final state of all metrics to a file of the user's choosing. It can do this in test or production environments. It doesn't perform any assertions, or otherwise do things that only make sense in a testing context. It is fairly similar to [Slf4jReporter](https://github.com/apache/flink/blob/release-1.5.5/flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java#L41), but its output is easier to parse (for machines or humans), and would normally be configured like:

```
metrics.reporters: file,
metrics.reporter.file.class: org.apache.beam.runners.flink.metrics.FileReporter,
metrics.reporter.file.file: /some/path/on/job-mgr/node
```

To me, it feels weirder to name it `Test*` and leave it under `main/`, when it is not really *that* test-specific an abstraction.

However I understand if the possible non-test uses feel contrived, and perhaps having `Test*` names under `main/` isn't too smelly.

I just want to make sure the vestigial `metrics.reporter.test.*` namespacing isn't driving the decision too heavily.

Reply:
That sounds reasonable. But shouldn't we use a more distinct name to avoid confusion, like `BeamFlinkFileMetricReporter`?

Reply:
I suppose it could come in handy when you just want to dump metrics locally.

Since package is `org.apache.beam.runners.flink.metrics`, `FileReporter` seems fine.

Reply:
fwiw: "beam", "flink", "metrics", "file", and "reporter" are already in this class' name: `org.apache.beam.runners.flink.metrics.FileReporter` (and users would use this FQN in `flink-conf.yaml`, as in the example above).

I generally dislike the Java convention of repeating package-name segments in class' basenames, but recognize that it's common and people seem to like it, so happy to add here, just wanted to float the idea that `org.apache.beam.runners.flink.metrics.BeamFlinkFileMetricReporter` is redundant.

lmk if you still prefer it and I'll change it.

Reply:
Reason I brought it up is discoverability. Names like `Node`, `Metric`, `Reporter` all usually too generic, regardless in which package they reside. When you work in the IDE, you would probably not want to see dozens of matches when searching a class? I checked and in this case it is not the case however and not critical for this PR anyways. BTW it seems to be a convention in Beam code to prefix classes with the runner (like `Flink` when they are likely to appear in multiple runners).